src/table.rs

```diff
@@ -338,6 +334,124 @@ where
+    ///
+    /// ```rust
+    /// use tokengrams::SuffixTable;
+    /// use utf16_literal::utf16;
+    ///
+    /// let sa = SuffixTable::new(utf16!("The quick brown fox.").to_vec(), None, false);
+    /// assert!(sa.contains(utf16!("quick")));
+    /// ```
+    #[allow(dead_code)]
+    fn contains(&self, query: &[usize]) -> bool;
+
+    /// Returns an unordered list of positions where `query` starts in `text`.
+    ///
+    /// This runs in `O(mlogn)` time, where `m == query.len()` and
+    /// `n == self.len()`. (As far as this author knows, this is the best known
+    /// bound for a plain suffix table.)
+    ///
+    /// Positions are byte indices into `text`.
+    ///
+    /// If you just need to test existence, then use `contains` since it is
+    /// faster.
+    ///
+    /// # Example
+    ///
+    /// Build a suffix array of some text and find all occurrences of a
+    /// substring:
+    ///
+    /// ```rust
+    /// use tokengrams::SuffixTable;
+    /// use utf16_literal::utf16;
+    ///
+    /// let sa = SuffixTable::new(utf16!("The quick brown fox was very quick.").to_vec(), None, false);
+    /// assert_eq!(sa.positions(utf16!("quick")), &[4, 29]);
+    /// ```
+    #[allow(dead_code)]
+    fn positions(&self, query: &[usize]) -> &[u64];
+
+    // Count occurrences of each token directly following the query sequence.
+    fn count_next(&self, query: &[usize]) -> Vec<usize>;
+
+    fn batch_count_next(&self, queries: &[Vec<usize>]) -> Vec<Vec<usize>>;
+
+    // For a given n, produce a map from an occurrence count to the number of unique n-grams with that occurrence count.
+    fn count_ngrams(&self, n: usize) -> HashMap<usize, usize>;
+}
+
+#[typetag::serialize]
+pub trait InMemoryTable: Table {}
+
+#[typetag::serialize]
+impl InMemoryTable for SuffixTable<Box<[u16]>> {}
+
+#[typetag::serialize]
+impl InMemoryTable for SuffixTable<Box<[u32]>> {}
+
+impl<T, U, E> Table for SuffixTable<T, U>
+where
+    T: Deref<Target = [E]> + Sync,
+    U: Deref<Target = [u64]> + Sync,
+    E: Unsigned,
+{
+    fn is_sorted(&self) -> bool {
+        self.is_sorted()
+    }
+
+    fn contains(&self, query: &[usize]) -> bool {
+        let query: Vec<E> = query.iter()
+            .filter_map(|&item| E::try_from(item).ok())
+            .collect();
+        self.contains(&query)
+    }
+
+    fn positions(&self, query: &[usize]) -> &[u64] {
+        let query: Vec<E> = query.iter()
+            .filter_map(|&item| E::try_from(item).ok())
+            .collect();
+        self.positions(&query)
+    }
+
+    fn count_next(&self, query: &[usize]) -> Vec<usize> {
+        let query: Vec<E> = query.iter()
+            .filter_map(|&item| E::try_from(item).ok())
+            .collect();
+        self.count_next(&query)
+    }
+
+    fn batch_count_next(&self, queries: &[Vec<usize>]) -> Vec<Vec<usize>> {
+        queries
+            .into_par_iter()
+            .map(|query| {
+                let query = unsafe { std::mem::transmute(query.as_slice()) };
```

Comment:
Wait why do we need to do an unsafe transmute here? Wouldn't `as_slice` suffice?

Reply:
ahh sorry I thought this was gone, let me get rid of it